File: internal/runtime/active_run.go
package runtime

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"
)

type ResolvedRunDir struct {
	RunDir  string
	Source  string
	Created bool
	Label   *string
	Root    *string
}

func ReadActiveRunDir(env map[string]string) (*string, error) {
	fromEnv := env[ActiveRunEnvVar]
	if stringsTrim(fromEnv) == "" {
		return nil, nil
	}
	resolved, err := filepathAbs(fromEnv)
	if err != nil {
		return nil, err
	}
	if err := mustDir(resolved, ActiveRunEnvVar); err != nil {
		return nil, err
	}
	return &resolved, nil
}

func ResolveRunDir(outputDir *string, root *string, label *string, env map[string]string, now time.Time, cwd string) (*ResolvedRunDir, error) {
	if outputDir != nil && stringsTrim(*outputDir) != "" {
		resolved := resolvePath(cwd, *outputDir)
		if err := os.MkdirAll(resolved, 0o755); err != nil {
			return nil, err
		}
		return &ResolvedRunDir{
			RunDir:  resolved,
			Source:  "explicit",
			Created: false,
		}, nil
	}
	if envValue := stringsTrim(env[ActiveRunEnvVar]); envValue != "" {
		active, err := ReadActiveRunDir(map[string]string{
			ActiveRunEnvVar: resolvePath(cwd, envValue),
		})
		if err != nil {
			return nil, err
		}
		if active == nil {
			return nil, fmt.Errorf("%s does not exist", ActiveRunEnvVar)
		}
		return &ResolvedRunDir{
			RunDir:  *active,
			Source:  "active",
			Created: false,
		}, nil
	}
	run, err := StartWorkspaceRun(root, label, now, cwd)
	if err != nil {
		return nil, err
	}
	rootValue := run.Root
	labelValue := run.Label
	return &ResolvedRunDir{
		RunDir:  run.RunDir,
		Source:  "auto",
		Created: true,
		Label:   &labelValue,
		Root:    &rootValue,
	}, nil
}

func stringsTrim(value string) string {
	return strings.TrimSpace(value)
}

func filepathAbs(path string) (string, error) {
	return filepath.Abs(path)
}

File: scripts/agent-runtime/active-run.mjs
import { existsSync, mkdirSync, statSync } from "node:fs";
import { resolve } from "node:path";
import process from "node:process";
import { createRun } from "./new-workspace-run.mjs";

export const ACTIVE_RUN_ENV_VAR = "CAUTILUS_RUN_DIR";
export const DEFAULT_RUNS_ROOT = ".cautilus/runs";

function ensureExistingDirectory(path, label) {
	if (!existsSync(path)) {
		throw new Error(`${label} does not exist: ${path}`);
	}
	if (!statSync(path).isDirectory()) {
		throw new Error(`${label} must be a directory: ${path}`);
	}
}

export function resolveRunDir({
	outputDir = null,
	root = null,
	label = null,
	env = process.env,
	now = new Date(),
	cwd = process.cwd(),
} = {}) {
	if (outputDir) {
		const resolved = resolve(cwd, outputDir);
		mkdirSync(resolved, { recursive: true });
		return {
			runDir: resolved,
			source: "explicit",
			created: false,
			label: null,
			root: null,
		};
	}

	const fromEnv = env?.[ACTIVE_RUN_ENV_VAR];
	if (fromEnv) {
		const resolved = resolve(cwd, fromEnv);
		ensureExistingDirectory(resolved, ACTIVE_RUN_ENV_VAR);
		return {
			runDir: resolved,
			source: "active",
			created: false,
			label: null,
			root: null,
		};
	}

	const rootPath = resolve(cwd, root ?? DEFAULT_RUNS_ROOT);
	mkdirSync(rootPath, { recursive: true });
	const result = createRun({ root: rootPath, label, now });
	return {
		runDir: result.runDir,
		source: "auto",
		created: true,
		label: result.label,
		root: rootPath,
	};
}
